tests/mocha/block_test.js

```diff
@@ -56,142 +45,285 @@ suite('Blocks', function() {
+          blocks.B.unplug(true);
+          // Even though we're asking to heal, it will appear as if it has not
+          // healed because shadows always stay with the parent.
+          assertUnpluggedNoheal(blocks);
+        });
       });
-      test('C is Shadow', function() {
-        var blocks = this.blocks;
-        blocks.C.setShadow(true);
-        blocks.B.unplug(true);
-        // Even though we're asking to heal, it will appear as if it has not
-        // healed because shadows always stay with the parent.
-        assertUnpluggedNoheal(blocks);
+      suite('Stack', function() {
+        setup(function() {
+          var blockA = this.workspace.newBlock('stack_block');
+          var blockB = this.workspace.newBlock('stack_block');
+          var blockC = this.workspace.newBlock('stack_block');
+
+          blockA.nextConnection.connect(blockB.previousConnection);
+          blockB.nextConnection.connect(blockC.previousConnection);
+
+          assertEquals(blockB, blockC.getParent());
+
+          this.blocks = {
+            A: blockA,
+            B: blockB,
+            C: blockC
+          };
+        });
+
+        test('Don\'t heal', function() {
+          this.blocks.B.unplug();
+          assertUnpluggedNoheal(this.blocks);
+        });
+        test('Heal', function() {
+          this.blocks.B.unplug(true);
+          assertUnpluggedHealed(this.blocks);
+        });
+        test.skip('Heal with bad checks', function() {
+          var blocks = this.blocks;
+          // A and C can't connect, but both can connect to B.
+          blocks.A.nextConnection.setCheck('type1');
+          blocks.C.previousConnection.setCheck('type2');
+
+          // The types don't work.
+          blocks.B.unplug(true);
+
+          // TODO (#1994): Check types before unplugging. Currently
+          //  everything disconnects, when C should stick with B
+          assertUnpluggedNoheal();
+        });
+        test('C is Shadow', function() {
+          var blocks = this.blocks;
+          blocks.C.setShadow(true);
+          blocks.B.unplug(true);
+          // Even though we're asking to heal, it will appear as if it has not
+          // healed because shadows always stay with the parent.
+          assertUnpluggedNoheal(blocks);
+        });
       });
     });
-    suite('Stack', function() {
-      setup(function() {
-        var blockA = this.workspace.newBlock('stack_block');
-        var blockB = this.workspace.newBlock('stack_block');
-        var blockC = this.workspace.newBlock('stack_block');
-
-        blockA.nextConnection.connect(blockB.previousConnection);
-        blockB.nextConnection.connect(blockC.previousConnection);
-
-        assertEquals(blockB, blockC.getParent());
-
-        this.blocks = {
-          A: blockA,
-          B: blockB,
-          C: blockC
-        };
-      });
+    suite('Dispose', function() {
+      function assertDisposedNoheal(blocks) {
+        chai.assert.isNotOk(blocks.A.disposed);
```

Comment:
Why check `isNotOk` if you want it to be `false`, but `isTrue` if you want it to be `true`?  

Is it because blocks don't have a `disposed` property until they've been disposed of?  If so, why not have that property (`Blockly.Block.prototype.disposed = false`) and then set it to `true` on the block when dispose is called?

Edit: and if the answer is that you'll get to it when you get to block disposing, that's fine too.

Reply:
> Is it because blocks don't have a `disposed` property until they've been disposed of?

You are correct!

> If so, why not have that property (`Blockly.Block.prototype.disposed = false`) and then set it to `true` on the block when dispose is called?

Sounds good, I'll go ahead and add it, plus update the connections and fields as well. First thing tomorrow :D

Reply:
Fixed!